activation/handler_v2.go

```diff
@@ -248,3 +306,319 @@ func (h *HandlerV2) collectAtxDeps(atx *wire.ActivationTxV2) ([]types.Hash32, []
+	var minLeaves uint64
+	var smesherCommitment *types.ATXID
+	for _, niposts := range atx.NiPosts {
+		// verify PoET memberships in a single go
+		var poetChallenges [][]byte
+
+		for _, post := range niposts.Posts {
+			if post.MarriageIndex >= uint32(len(equivocationSet)) {
+				err := fmt.Errorf("marriage index out of bounds: %d > %d", post.MarriageIndex, len(equivocationSet)-1)
+				return 0, 0, nil, err
+			}
+			id := equivocationSet[post.MarriageIndex]
+			effectiveNumUnits := post.NumUnits
+			if atx.Initial == nil {
+				var err error
+				effectiveNumUnits, err = h.validatePreviousAtx(id, &post, previousAtxs)
+				if err != nil {
+					return 0, 0, nil, fmt.Errorf("validating previous atx for ID %s: %w", id, err)
+				}
+			}
+			totalEffectiveNumUnits += effectiveNumUnits
+
+			var commitment types.ATXID
+			if atx.Initial != nil {
+				commitment = atx.Initial.CommitmentATX
+			} else {
+				var err error
+				commitment, err = atxs.CommitmentATX(h.cdb, id)
+				if err != nil {
+					return 0, 0, nil, fmt.Errorf("commitment atx not found for ID %s: %w", id, err)
+				}
+				if smesherCommitment == nil {
+					smesherCommitment = &commitment
+				}
+			}
+
+			err := h.nipostValidator.PostV2(
+				ctx,
+				id,
+				commitment,
+				&post.Post,
+				niposts.Challenge[:],
+				post.NumUnits,
+				PostSubset([]byte(h.local)),
+			)
+			if err != nil {
+				// TODO generate malfeasance proof
+				return 0, 0, nil, fmt.Errorf("invalid post for ID %s: %w", id, err)
+			}
+
+			nipostChallenge := wire.NIPostChallengeV2{
+				PublishEpoch:     atx.PublishEpoch,
+				PositioningATXID: atx.PositioningATX,
+			}
+			if atx.Initial != nil {
+				nipostChallenge.InitialPost = &atx.Initial.Post
+			} else {
+				nipostChallenge.PrevATXID = atx.PreviousATXs[post.PrevATXIndex]
+			}
+
+			poetChallenges = append(poetChallenges, nipostChallenge.Hash().Bytes())
+		}
+		membership := types.MultiMerkleProof{
+			Nodes:       niposts.Membership.Nodes,
+			LeafIndices: niposts.Membership.LeafIndices,
+		}
+		leaves, err := h.nipostValidator.PoetMembership(ctx, &membership, niposts.Challenge, poetChallenges)
+		if err != nil {
+			return 0, 0, nil, fmt.Errorf("invalid poet membership: %w", err)
+		}
+		minLeaves = min(leaves, minLeaves)
+	}
+
+	if atx.Initial == nil {
+		if err := h.validateVrfNonce(atx, previousAtxs[0], *smesherCommitment); err != nil {
+			return 0, 0, nil, fmt.Errorf("validating VRF nonce: %w", err)
+		}
+	}
+
+	return minLeaves, totalEffectiveNumUnits, nil, nil
+}
+
+func (h *HandlerV2) checkMalicious(
+	ctx context.Context,
+	tx *sql.Tx,
+	watx *wire.ActivationTxV2,
+) (bool, *mwire.MalfeasanceProof, error) {
+	malicious, err := identities.IsMalicious(tx, watx.SmesherID)
+	if err != nil {
+		return false, nil, fmt.Errorf("checking if node is malicious: %w", err)
+	}
+	if malicious {
+		return true, nil, nil
+	}
+
+	// TODO:
+	// 1. check double-publish
+	// 2. check previous ATX
+	return false, nil, nil
```

Comment:
and 

- same node ID in multiple marriage certificates
- two ATXs referencing the same marriage certificate in the same epoch (we probably need a DB migration for that to do this lookup efficiently?)
-  an ATX without marriage certificate contains a post of an identity that also produced a post for one with a marriage certificate